src/lib/logical_query_plan/enable_make_for_lqp_node.hpp

```diff
@@ -10,68 +10,75 @@ namespace opossum {
+          if constexpr (IsLQPNodeArgument<sizeof...(ArgumentTypes) - 2, ArgumentTypes...>::value) {
+            // Use function arguments, except for the last two, to construct the LQP node.
+            auto node = create_lqp_node(arguments_tuple, std::make_index_sequence<sizeof...(ArgumentTypes) - 2>());
+            node->set_left_input(std::get<sizeof...(ArgumentTypes) - 2>(arguments_tuple));
+            node->set_right_input(std::get<sizeof...(ArgumentTypes) - 1>(arguments_tuple));
             return node;
           } else {
-            // last argument is shared_ptr<AbstractLQPNode>
-            auto node = make_impl(args_tuple, std::make_index_sequence<sizeof...(Args)-1>());
-            node->set_left_input(std::get<sizeof...(Args)-1>(args_tuple));
+            // Use function arguments, except for the last, to construct the LQP node.
+            auto node = create_lqp_node(arguments_tuple, std::make_index_sequence<sizeof...(ArgumentTypes) - 1>());
+            node->set_left_input(std::get<sizeof...(ArgumentTypes) - 1>(arguments_tuple));
             return node;
           }
         } else {
-          // last argument is shared_ptr<AbstractLQPNode>
-          auto node = make_impl(args_tuple, std::make_index_sequence<sizeof...(Args)-1>());
-          node->set_left_input(std::get<sizeof...(Args)-1>(args_tuple));
+          // Only one LQP input node was provided as an argument.
+          auto node = std::make_shared<DerivedNode>();
+          node->set_left_input(std::get<0>(arguments_tuple));
           return node;
         }
       } else {
-        // no shared_ptr<AbstractLQPNode> was passed at the end
-        return make_impl(std::forward_as_tuple(args...), std::make_index_sequence<sizeof...(Args)-0>());
+        // Additional LQP input nodes were not passed as last arguments.
+        return create_lqp_node(arguments_tuple, std::make_index_sequence<sizeof...(ArgumentTypes)>());
       }
     } else {
-      // no shared_ptr<AbstractLQPNode> was passed at the end
-      return make_impl(std::forward_as_tuple(args...), std::make_index_sequence<sizeof...(Args)-0>());
+      // No arguments were passed to this function.
+      return std::make_shared<DerivedNode>();
     }
-    // clang-format on
   }
 
  private:
-  template <class Tuple, size_t... I>
-  static std::shared_ptr<DerivedNode> make_impl(const Tuple& constructor_arguments,
-                                                std::index_sequence<I...> num_constructor_args) {
-    return std::make_shared<DerivedNode>(std::get<I>(constructor_arguments)...);
+  template <class ArgumentsTupleType, size_t... ConstructorIndices>
+  static std::shared_ptr<DerivedNode> create_lqp_node(const ArgumentsTupleType& arguments_tuple,
+                                                      std::index_sequence<ConstructorIndices...> /* indices */) {
+    return std::make_shared<DerivedNode>(std::get<ConstructorIndices>(arguments_tuple)...);
```

Comment:
I don't get it. I haven't found a std::get version that accepts a sequence. How does that exactly work here?

Reply:
True, this compiler magic is not very intuitive. My understanding:

* `ConstructorIndices` represents a non-type template parameter pack. For example the values `0, 1, 2`
* The `...` in `std::get<ConstructorIndices>(arguments_tuple)...` triggers parameter pack expansion, which is described as follows: 

  > A pattern followed by an ellipsis, in which the name of at least one parameter pack appears at least once, is expanded into zero or more comma-separated instantiations of the pattern, where the name of the parameter pack is replaced by each of the elements from the pack, in order